File: backend/app/schemas/transcript_api_ingestion.py
"""Schemas for API-based transcript ingestion queue."""

from datetime import datetime

from pydantic import BaseModel, Field


class CatalogCoverageStats(BaseModel):
    """Catalog-wide transcript / embedding coverage (not run-specific)."""

    total_videos: int
    with_transcript: int
    without_transcript: int
    with_transcript_embedding: int
    transcript_coverage_pct: float = 0.0
    embedding_coverage_pct: float = 0.0


class RunJobStats(BaseModel):
    """Per-run job counters and derived progress."""

    jobs_total: int
    queued: int = 0
    processing: int = 0
    success: int = 0
    failed: int = 0
    unavailable: int = 0
    skipped_existing: int = 0
    processed: int = 0
    progress_pct: float = 0.0
    success_pct: float = 0.0
    unavailable_pct: float = 0.0


class TranscriptApiIngestionDashboardStats(BaseModel):
    """GET /stats — catalog only + pointer to active run."""

    catalog: CatalogCoverageStats
    active_run_id: int | None = None


# Backward-compatible alias for older clients
class TranscriptApiIngestionStats(CatalogCoverageStats):
    """Deprecated merged shape; prefer CatalogCoverageStats + RunJobStats."""

    queued: int = 0
    processing: int = 0
    success: int = 0
    failed: int = 0
    unavailable: int = 0
    skipped_existing: int = 0


class TranscriptApiIngestionStartRequest(BaseModel):
    limit: int = Field(default=500, ge=1, le=5000)
    worker_count: int = Field(default=5, ge=1, le=10)
    creator_filter: str | None = Field(default=None, max_length=255)
    latest_only: bool = False
    only_missing: bool = True


class TranscriptApiIngestionJobRead(BaseModel):
    id: int
    run_id: int
    video_id: int
    status: str
    title: str
    creator_name: str
    transcript_chars: int
    embedding_created: bool
    sheets_rows_updated: int
    sheets_writeback: str
    error_message: str | None
    updated_at: datetime

    model_config = {"from_attributes": True}


class TranscriptApiIngestionJobsPage(BaseModel):
    items: list[TranscriptApiIngestionJobRead]
    total: int
    offset: int
    limit: int
    status_filter: str | None = None


class TranscriptApiIngestionRunRead(BaseModel):
    id: int
    status: str
    worker_count: int
    limit_count: int | None
    creator_filter: str | None
    latest_only: bool
    only_missing: bool
    jobs_total: int
    message: str | None
    error_message: str | None
    started_at: datetime
    finished_at: datetime | None
    duration_seconds: int | None
    catalog: CatalogCoverageStats
    run_stats: RunJobStats
    jobs: list[TranscriptApiIngestionJobRead] = Field(default_factory=list)

    model_config = {"from_attributes": True}


class TranscriptApiIngestionStartResponse(BaseModel):
    run: TranscriptApiIngestionRunRead
    message: str

File: frontend/types/transcript-api-ingestion.ts
/** API-based transcript ingestion queue (server-side, no browser). */

export type TranscriptApiIngestionJobStatus =
  | "queued"
  | "processing"
  | "success"
  | "failed"
  | "unavailable"
  | "skipped_existing";

export type TranscriptApiIngestionRunStatus =
  | "queued"
  | "running"
  | "paused"
  | "completed"
  | "failed";

export type JobStatusFilter = "all" | TranscriptApiIngestionJobStatus;

export interface CatalogCoverageStats {
  total_videos: number;
  with_transcript: number;
  without_transcript: number;
  with_transcript_embedding: number;
  transcript_coverage_pct: number;
  embedding_coverage_pct: number;
}

export interface RunJobStats {
  jobs_total: number;
  queued: number;
  processing: number;
  success: number;
  failed: number;
  unavailable: number;
  skipped_existing: number;
  processed: number;
  progress_pct: number;
  success_pct: number;
  unavailable_pct: number;
}

export interface TranscriptApiIngestionDashboardStats {
  catalog: CatalogCoverageStats;
  active_run_id: number | null;
}

export interface TranscriptApiIngestionStartRequest {
  limit: number;
  worker_count: number;
  creator_filter?: string | null;
  latest_only: boolean;
  only_missing: boolean;
}

export interface TranscriptApiIngestionJob {
  id: number;
  run_id: number;
  video_id: number;
  status: TranscriptApiIngestionJobStatus;
  title: string;
  creator_name: string;
  transcript_chars: number;
  embedding_created: boolean;
  sheets_rows_updated: number;
  sheets_writeback: string;
  error_message: string | null;
  updated_at: string;
}

export interface TranscriptApiIngestionJobsPage {
  items: TranscriptApiIngestionJob[];
  total: number;
  offset: number;
  limit: number;
  status_filter: string | null;
}

export interface TranscriptApiIngestionRun {
  id: number;
  status: TranscriptApiIngestionRunStatus;
  worker_count: number;
  limit_count: number | null;
  creator_filter: string | null;
  latest_only: boolean;
  only_missing: boolean;
  jobs_total: number;
  message: string | null;
  error_message: string | null;
  started_at: string;
  finished_at: string | null;
  duration_seconds: number | null;
  catalog: CatalogCoverageStats;
  run_stats: RunJobStats;
  jobs: TranscriptApiIngestionJob[];
}

export interface TranscriptApiIngestionStartResponse {
  run: TranscriptApiIngestionRun;
  message: string;
}

export function isApiIngestionRunActive(
  run: TranscriptApiIngestionRun | null | undefined,
): boolean {
  if (!run) return false;
  return run.status === "queued" || run.status === "running" || run.status === "paused";
}

export function isApiIngestionRunTerminal(
  run: TranscriptApiIngestionRun | null | undefined,
): boolean {
  if (!run) return false;
  return run.status === "completed" || run.status === "failed";
}
